packages/flutter/test/widgets/navigator_test.dart

```diff
@@ -1930,6 +1930,77 @@ void main() {
     expect(popNextOfFirst, secondRoute);
   });
 
+  testWidgets('hero controller scope works', (WidgetTester tester) async {
+    final GlobalKey<NavigatorState> top = GlobalKey<NavigatorState>();
+    final GlobalKey<NavigatorState> sub = GlobalKey<NavigatorState>();
+
+    final List<NavigatorObservation> observations = <NavigatorObservation>[];
+    final HeroControllerSpy spy = HeroControllerSpy()
+      ..onPushed = (Route<dynamic> route, Route<dynamic> previousRoute) {
+        observations.add(
+          NavigatorObservation(
+            current: route?.settings?.name,
+            previous: previousRoute?.settings?.name,
+            operation: 'didPush'
+          )
+        );
+      };
+    await tester.pumpWidget(
+      HeroControllerScope(
+        controller: spy,
+        child: Directionality(
+          textDirection: TextDirection.ltr,
+          child: Navigator(
+            key: top,
+            initialRoute: 'top1',
+            onGenerateRoute: (RouteSettings s) {
+              return MaterialPageRoute<void>(
+                builder: (BuildContext c) {
+                  return Navigator(
+                    key: sub,
+                    initialRoute: 'sub1',
+                    onGenerateRoute: (RouteSettings s) {
+                      return MaterialPageRoute<void>(
+                        builder: (BuildContext c) {
+                          return const Placeholder();
+                        },
+                        settings: s,
+                      );
+                    },
+                  );
+                },
+                settings: s,
+              );
+            },
+          ),
+        )
+      )
+    );
+    // It should only observe the top navigator.
+    expect(observations.length, 1);
+    expect(observations[0].current, 'top1');
+    expect(observations[0].previous, isNull);
+
+    sub.currentState.push(MaterialPageRoute<void>(
+      settings: const RouteSettings(name:'sub2'),
+      builder: (BuildContext context) => const Text('sub2')
+    ));
+    await tester.pumpAndSettle();
+
+    expect(find.text('sub2'), findsOneWidget);
+    // It should not record sub navigator.
+    expect(observations.length, 1);
+
+    top.currentState.push(MaterialPageRoute<void>(
+      settings: const RouteSettings(name:'top2'),
+      builder: (BuildContext context) => const Text('top2')
+    ));
+    await tester.pumpAndSettle();
+    expect(observations.length, 2);
+    expect(observations[1].current, 'top2');
+    expect(observations[1].previous, 'top1');
+  });
+
```

Comment:
we should have a test that verifies that changing the hero controller mid-way through a hero animation works.